### DISCUSSION TITLE
Building Firefox with msys2 - Really?

### DISCUSSION THREAD

**Discussion by Biswa96** [msys2/MINGW-packages#10526] [General]
I have came across this tweet https://twitter.com/FirefoxCommit/status/1478984734038441987 and hence this Bugzilla report https://bugzilla.mozilla.org/show_bug.cgi?id=1725895. Is it really possible to build Firefox with msys2 and mingw toolchain?

There was a little hint in this issue https://github.com/msys2/MSYS2-packages/issues/2784. @mitchhentges Please forgive me for my curiosity. Is there any way msys2 project can help to make that reality or make the process smooth?

Firefox has many `#ifdef __MINGW32__` like workarounds in source code. Does mingw-w64 project also require any modification?

**Comment by revelator** [comment]
if i remember correctly it needs specially setup msys or msys2 env for use with msvc compiler (maybe that has changed been a while).

**Reply by revelator** [reply]
hmm seems i was right https://bugzilla.mozilla.org/show_bug.cgi?id=1100925 some discussion there and yes it seems msys2 needs to be setup for msvc interaction.

**Comment by mitchhentges** [comment]
Good questions, yeah, there's two contexts where MSYS2 are related to Firefox:

1. Compiling with the MinGW/GNU toolchain. I believe Firefox has been _possible_ to build with MinGW for some time, and [we have a CI job to ensure continued compatibility](https://treeherder.mozilla.org/jobs?repo=mozilla-central&searchStr=Windows%2CMinGW%2Call%2CMinGW-Clang%2Cbuilds%2Cfor%2CWindows%2C32-bits%2Cbuild-win32-mingwclang%2Fopt%2CBo&selectedTaskRun=D18Q0YbzS2G8rxM1zCfnUg.0). However, yeah, it might be only 32-bit MinGW supported, so I'm not sure how compatible Firefox is with the modern MinGW with MSYS2. On the flip side, I'm not sure what the advantage is of compiling Firefox with MinGW instead of MSVC - I suppose that you're able to use tools from the MinGW ecosystem, which can be better than MSVC in some cases? Unsure :)
2. There's some specific tools/binaries that the Firefox build system needs: `sh` (for parsing build config, which is weird!), `awk`, `tar` I believe, and some others. Since they're tricky to have on Windows, we've traditionally (since ~2006) shipped an MSYS-based development environment that developers would have to build from: this environment is the "MozillaBuild" project referred to in [the bug revelator linked](https://bugzilla.mozilla.org/show_bug.cgi?id=1100925).

The recent work I've been doing in the area (including the work [the tweet](https://twitter.com/FirefoxCommit/status/1478984734038441987) was posting about) was porting the development environment from being based on MSYS (1) to MSYS2. That migration has been surprisingly smooth, and the [few hiccups I've run into](https://github.com/msys2/MSYS2-packages/issues?q=is%3Aissue+author%3Amitchhentges+) have had really fast, helpful replies (shoutout to lazka and 1480c1, thanks for everything!).

So, TL;DR :)
* It's possible to build Firefox with "old" MinGW, unsure about current MSYS2 status
* The process of porting MozillaBuild from MSYS (1) to MSYS2 has been great, good work team! 👏 🎉

**Comment by wszqkzqk** [comment]
Is there any new progress? I think the biggest advantage is that it makes Firefox on Windows independent of proprietary compilers and makes it FOSS, allowing wine-gecko to be integrated into the software libraries of some strictly required Linux distributions. 😊

**Reply by Biswa96** [reply]
Guess what! Tor Browser was being built with mingw toolchain this entire time. I was notified by Tom Ritter, one of the Firefox developers. See this thread https://groups.google.com/a/mozilla.org/g/dev-platform/c/w8F43mKLmj0/m/zmZ0RUOcBwAJ

But it seems that the road is bumpy. Hang tight, developers!

**Reply by revelator** [reply]
or pale moon maybe ?.

**Reply by Kreijstal** [reply]
pale moon uses python 2, unless you package tauthon it is going to be hard to build

**Reply by revelator** [reply]
ah well would have been nice...
qtweb still works mostly fine, the problem with it is that it relies on qt4 and an older model of qtwebkit, i tried to port it over to qt5 but the new webkit model is simply not compatible it seems :( otherwise it would have been a nice browser to include. 
pale moon still uses XUL while recent firefox seem to have changed to something else. i wonder if the python2 code could be ported to python3 ?.
